### PR TITLE
Allow plugins to define their own targets

### PR DESCRIPTION
**Pull request by AdrienClairembault**
Allow plugin to add additional targets.
This is done through a `add_formcreator_targets` hook like this:

```php
$PLUGIN_HOOKS['add_formcreator_targets']['plugin_demo'] = [
      EmailTarget::class
];
```

### CODE REVIEW

File: inc/hooksmigration.class.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      $query = "CREATE TABLE $table (
+         `id` int(11) NOT NULL auto_increment,
+         `name` varchar(255) collate utf8_unicode_ci default NULL,
+         `plugin_formcreator_forms_id` int(11) NOT NULL default '0',
+         `show_rule` int(11) NOT NULL DEFAULT '1',
+         PRIMARY KEY (`id`)
+      ) ENGINE=InnoDB  DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci";
+
+      $DB->queryOrDie($query, "Can't create $table: ". $DB->error());
+   }
```
Comment (btry) [inline inc/hooksmigration.class.php:51]: This create table statmeent must be in the plugin_formcreator_empty.sql file with a IF NOT EXIST condition.

This way, upgrade will handle it.

**Comment by AdrienClairembault** [comment]
Rebased and updated the description.
Should be OK to review now :)

File: inc/hooksmigration.class.php (hunk 1 above)
Comment (AdrienClairembault) [reply inc/hooksmigration.class.php:51]: No tables were added, this is only an helper method to be used by others plugins.
This allows them to easily create the 'base' target table in one line of code.

They can then add their specific fields manually.

**Comment by AdrienClairembault** [comment]
Up

**Review by btry — CHANGES_REQUESTED** [review]

File: inc/abstracttarget.class.php [hunk 2]
```diff
@@ -2593,16 +2559,5 @@ public function getTargetsForForm($formId) {
...
-      $item->initForm($item->getID(), $options);
-      $options['candel'] = false;
-      $options['formoptions'] = sprintf('data-itemtype="%s" data-id="%s"', self::getType(), $item->getID());
-      TemplateRenderer::getInstance()->display('@formcreator/pages/condition_for_item.html.twig', [
-         'item'   => $item,
-         'params' => $options,
-         'parent' => $item,
-      ]);
-      return true;
-   }
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply inc/abstracttarget.class.php:2607]: I need to use this method in my plugin.

```php
public static function displayTabContentForItem(CommonGLPI $item, $tabnum = 1, $withtemplate = 0) {
    switch ($item->getType()) {
        case self::class:
            /** @var EmailTarget $item */
            switch ($tabnum) {
                case 1:
                    self::showConditions($item);
                    break;
            }
            break;
    }

    return false;
}
```

How will it be done with twig ?

File: inc/abstracttarget.class.php [hunk 3]
```diff
@@ -286,19 +286,6 @@ public static function getEnumLocationRule() {
...
-    *
-    * @return array
-    */
-   public static function getEnumShowRule(): array {
-      return [
-       PluginFormcreatorCondition::SHOW_RULE_ALWAYS => __('Always generated', 'formcreator'),
-       PluginFormcreatorCondition::SHOW_RULE_HIDDEN => __('Disabled unless', 'formcreator'),
-       PluginFormcreatorCondition::SHOW_RULE_SHOWN  => __('Generated unless', 'formcreator'),
-      ];
-   }
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply inc/abstracttarget.class.php:300]: How can I share this code with an item that is a `PluginFormcreatorTargetInterface` but not an 
`AbstractTarget ` ?

File: inc/form.class.php [hunk 4]
```diff
@@ -1981,11 +1982,19 @@ public function getFields() : array {
...
-      return [
+      global $PLUGIN_HOOKS;
+
+      $targets = [
          PluginFormcreatorTargetTicket::class,
          PluginFormcreatorTargetChange::class,
          PluginFormcreatorTargetProblem::class,
       ];
+
+      foreach ($PLUGIN_HOOKS['add_formcreator_targets'] ?? [] as $plugin_targets) {
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply inc/form.class.php:1993]: Indeed, I wanted to rename it but forgot.
Done b7022860.

File: inc/abstracttarget.class.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (btry) [reply inc/abstracttarget.class.php:300]: If you have such need we may solve it by creating an intermediate abstract class. Something like this.

CommonDBTM
|- PluginFormcreatorAbstractTarget implements PluginFormcreatorTargetInterface
    |- PluginFormcreatorAbstractItilTarget
    |   |- PluginFormcreatorTargetTicket
    |   |- PluginFormcreatorTargetChange
    |   |- PluginFormcreatorTargetProblem
    |- PluginFooTargetBar

We will probably need to split the methods of PluginFormcreatorAbstractTarget to move some in PluginFormcreatorAbstractItilTarget.

File: inc/abstracttarget.class.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (btry) [reply inc/abstracttarget.class.php:2607]: Like this

```twig
                    {% set conditions = call('PluginFormcreatorCondition::getConditionsFromItem', [item]) %}
                    {% for condition in conditions %}
                        {{ include('@formcreator/components/form/condition.html.twig') }}
                    {% endfor %}
```

File: ajax/condition.php [hunk 5]
```diff
@@ -32,6 +32,11 @@
 include ('../../../inc/includes.php');
 Session::checkRight('entity', UPDATE);
 
+// Clear doubled "\" in namespaces
+if (isset($_POST['itemtype'])) {
+   $_POST['itemtype'] = str_replace("\\\\", "\\", $_POST['itemtype']);
+}
+
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply ajax/condition.php]: No longer relevant, this code was removed since namespace were fixed in another core PR.

**Comment by AdrienClairembault** [comment]
I've rebased on #2647 to do some testing (so the diff is a bit broken for now).
Will be all good once #2647 is merged and I can rebase on develop

**Comment by btry** [comment]
Hi

#2647 : unit tests are green; it is waiting for your review.

**Comment by AdrienClairembault** [comment]
> Hi
> 
> #2647 : unit tests are green; it is waiting for your review.

Review was done this morning, you may have missed it ;)

File: inc/form.class.php [hunk 6]
```diff
@@ -461,7 +461,6 @@ public function showTargets($ID, $options = []) {
       echo '</tr>';
 
       $allTargets = $this->getTargetsFromForm();
-      $token = Session::getNewCSRFToken();
```
Comment (AdrienClairembault) [inline inc/form.class.php:464]: Seems like this token wasn't being used.

**Comment by AdrienClairembault** [comment]
Ready for review :)

File: js/scripts.js [hunk 7]
```diff
@@ -1326,22 +1326,22 @@ function plugin_formcreator_addTarget(items_id) {
    });
 }
 
-function plugin_formcreator_deleteTarget(itemtype, target_id) {
-   if(confirm(i18n.textdomain('formcreator').__('Are you sure you want to delete this destination:', 'formcreator'))) {
+$(document).on('click', '.formcreator_delete_target', function() {
```
Comment (btry) [inline js/scripts.js:1329]: It would be better to keep this as a function, but moving it in plugin_formcreator.

**Review by btry — APPROVED** [review]

**Pull request merged**